src/repr/src/row/collection.rs

```diff
@@ -0,0 +1,731 @@
+            } else {
+                *diff_idx += count;
+                count = 0;
+            }
+        }
+    }
+
+    /// Helper method for implementing [`RowIterator`].
+    ///
+    /// Projects columns for the provided `row`.
+    fn project<'a>(
+        projection: Option<&'a Vec<usize>>,
+        row: &'a RowRef,
+        datum_buf: &'a mut DatumVec,
+        row_buf: &'a mut Row,
+    ) -> Option<&'a RowRef> {
+        if let Some(projection) = projection.as_ref() {
+            // Copy the required columns into our reusable buffer.
+            {
+                let datums = datum_buf.borrow_with(row);
+                row_buf
+                    .packer()
+                    .extend(projection.iter().map(|i| &datums[*i]));
+            }
+
+            Some(row_buf)
+        } else {
+            Some(row)
+        }
+    }
+}
+
+impl RowIterator for SortedRowCollectionIter {
+    fn next(&mut self) -> Option<&RowRef> {
+        // Advance past our offset.
+        if let Some(offset) = self.offset {
+            Self::advance_by(
+                &self.collection,
+                &mut self.row_idx,
+                &mut self.diff_idx,
+                offset,
+            );
+        }
+        self.offset = None;
+
+        // Bail if we've reached our limit.
+        if let Some(0) = self.limit {
+            return None;
+        }
+
+        let row = self.collection.get(self.row_idx).map(|(r, _)| r)?;
+
+        // If we're about to yield a row, then subtract from our limit.
+        if let Some(limit) = &mut self.limit {
+            *limit = limit.saturating_sub(1);
+        }
+
+        // Advance to the next row.
+        Self::advance_by(&self.collection, &mut self.row_idx, &mut self.diff_idx, 1);
+
+        // Project away and/or re-order any columns.
+        let (datum_buf, row_buf) = &mut self.projection_buf;
+        let row = Self::project(self.projection.as_ref(), row, datum_buf, row_buf)?;
+
+        Some(row)
+    }
+
+    fn peek(&mut self) -> Option<&RowRef> {
+        // Even for peek we still have to advance past any offset, otherwise we
+        // could return a Row that shouldn't have been yielded.
+        if let Some(offset) = self.offset {
+            Self::advance_by(
+                &self.collection,
+                &mut self.row_idx,
+                &mut self.diff_idx,
+                offset,
+            );
+        }
+        self.offset = None;
+
+        // Bail if we've reached our limit.
+        if let Some(0) = self.limit {
+            return None;
+        }
+
+        let row = self.collection.get(self.row_idx).map(|(r, _)| r)?;
+
+        // Note: Unlike `next()` we do not subtract from our limit, nor advance
+        // the internal pointers.
+
+        // Project away and/or re-order any columns.
+        let (datum_buf, row_buf) = &mut self.projection_buf;
+        let row = Self::project(self.projection.as_ref(), row, datum_buf, row_buf)?;
+
+        Some(row)
+    }
```

Comment:
Can `next` be implemented in terms of `peek` to avoid the code duplication?

Reply:
Great idea!

Reply:
This is actually a bit trickier to do than I thought because of order in which we need to borrow things. Going to leave the code as-is if that's okay